deploy/buildtools/buildfarm.py

```diff
@@ -0,0 +1,334 @@
+    def wait_on_build_host_initialization(self, build_config: BuildConfig) -> None:
+        """Ensure build host is launched and ready to be used.
+
+        Args:
+            build_config: Build config used to find build host that must ready.
+        """
+        raise NotImplementedError
+
+    def get_build_host(self, build_config: BuildConfig) -> BuildHost:
+        """Get build host associated with the build config.
+
+        Args:
+            build_config: Build config used to find the build host.
+
+        Returns:
+            Build host associated with the build config.
+        """
+        for build_host in self.build_hosts:
+            if build_host.build_config == build_config:
+                return build_host
+
+        raise Exception(f"Unable to find build host for {build_config.name}")
+
+    def get_build_host_ip(self, build_config: BuildConfig) -> str:
+        """Get IP address associated with this dispatched build host.
+
+        Args:
+            build_config: Build config to find build host for.
+
+        Returns:
+            IP address for the specific build host.
+        """
+        return self.get_build_host(build_config).ip_address
+
+    @abc.abstractmethod
+    def release_build_host(self, build_config: BuildConfig) -> None:
+        """Release the build host.
+
+        Args:
+            build_config: Build config to find build host to terminate.
+        """
+        raise NotImplementedError
+
+@BuildFarm.register
+class IPAddrBuildFarm(BuildFarm):
+    """Build farm that selects from a set of user-determined IPs to allocate a new build host.
+
+    Attributes:
+        build_hosts_allocated: Count of build hosts assigned with builds (`BuildConfig`s).
+    """
+    build_hosts_allocated: int
+
+    def __init__(self, args: Dict[str, Any]) -> None:
+        """
+        Args:
+            args: Args (i.e. options) passed to the build farm.
+        """
+        BuildFarm.__init__(self, args)
+        self.build_hosts_allocated = 0
+
+    @staticmethod
+    def NAME() -> str:
+        """Human-readable name (used as the "build-farm-type" in the YAML).
+
+        Returns:
+            Human-readable name.
+        """
+        return "unmanaged"
+
+    def parse_args(self) -> None:
+        """Parse build host arguments."""
+        build_farm_hosts_key = "build-farm-hosts"
+        build_farm_hosts_list = self.args[build_farm_hosts_key]
+
+        default_build_dir = self.args["default-build-dir"]
+
+        # allocate N build hosts
+        for build_farm_host in build_farm_hosts_list:
+            build_farm_host_alloc = BuildHost()
+            if type(build_farm_host) is dict:
+                # add element { ip-addr: { arg1: val1, arg2: val2, ... } }
+                assert(len(build_farm_host.keys()) == 1)
+
+                ip_addr = list(build_farm_host.keys())[0]
+                ip_args = list(build_farm_host.values())[0]
+
+                dest_build_dir = list(build_farm_host.keys())[0]
```

Comment:
Shouldn't this be something from inside `ip_args`?  This doesn't look correct.
```suggestion
                dest_build_dir = ip_args.get('override-build-dir', default_build_dir)
```
maybe?  It is getting later than I should be looking at code...

Reply:
Yup thats right. I mis-copied things. Like I said, I didn't do any reasonable amount of testing outside of running mypy